src/socket.rs

```diff
@@ -59,6 +65,19 @@ impl UnixSeqpacket {
 		Ok(socket)
 	}
 
+	/// Connect to a socket without checking for the writeable flag
```

Comment:
Thanks for the PR!

It can certainly be useful to do initialization before going into asynchronous behavior and then hand off the socket to a spawned task later. So I see a good use case for having a function like this.

But I don't think we would want to return an unusable socket. Especially since we provide no way to actually wait for the connect attempt to finish later. Instead, I think we should actually block on the `connect()` call instead. And then I would rename the call to `connect_blocking(...)` to more clearly indicate the intended use.

What do you think?

Reply:
I was thinking about the blocking `connect`, but not familiar with tokio / async rust enough to implement timeout. The solutions I think of would require removing `O_NONBLOCK` + threads or custom epoll + timer fd or [Handle::block_on](https://docs.rs/tokio/latest/tokio/runtime/struct.Handle.html#method.block_on) (this has foot gun though).

With `try_connect` I can just write a for loop until `connect` succeeded. Although the semantics are different from timeout, this works good enough for most use cases.

Reply:
It's true that a non-blocking connect needs some finicking to get to work. But we could create the socket in blocking mode, then connect, and then enable non-blocking mode.

> With `try_connect` I can just write a for loop until `connect` succeeded (EDIT: ooops, the code should return Err even for EWOULDBLOCK in try_connect). Although the semantics are different from timeout, this works good enough for most use cases.

I don't think this really works here. This function creates a new socket, so calling it in a loop doesn't really give the OS more time to set up the connection. It has to start from scratch each time.

In practice it still works because the connect almost certainly succeeds on the first call, since there is no actual networking involved.

Reply:
Wait you may be right about `try_connect` could return unusable socket. But I think the same thing also applies to calling connect on blocking fd, and we still need to check for writability. Because connect will return when the fd is added to backlog queue but if the listener never calls accept then it is unusable.

Reply:
We don't have to worry about the listener not calling `accept()`, that's really not on us as the connecting side. But we do have to care about `connect()` completing.

From the POSIX specification:
> If  the  initiating  socket is connection-mode, then connect() shall attempt to establish a connection to the address specified by the address argument. If the connection cannot be established immediately and O_NONBLOCK is not set for the file descriptor for the socket, connect() shall block for up to an unspecified timeout interval until the connection is established. If the timeout interval expires before the connection  is  established,  connect()  shall  fail and the connection attempt shall be aborted. If connect() is interrupted by a signal that is caught while blocked waiting to establish a connection, connect() shall fail and set errno  to [EINTR],  but  the  connection  request  shall  not be aborted, and the connection shall be established asynchronously.
>
> If the connection cannot be established immediately and O_NONBLOCK is set for the  file  descriptor  for  the socket, connect() shall fail and set errno to [EINPROGRESS], but the connection request shall not be aborted, and the connection shall be established asynchronously. Subsequent calls to connect() for  the  same  socket, before the connection is established, shall fail and set errno to [EALREADY].
>
> When  the connection has been established asynchronously, pselect(), select(), and poll() shall indicate that the file descriptor for the socket is ready for writing.

In other words: waiting for the file descriptor to become readable is just the kernel telling us that the earlier `connect()` attempt finished. And in non-blocking mode, `connect()` will just block until that has happened.

Whether or not this means that the listener called `accept()` doesn't really matter to us here. We're just following POSIX semantics.

Reply:
Thanks for the clarification. I see what you mean now. Let me fix up the PR.